pkg/reconciler/events/pubsub/pubsub.go

```diff
@@ -63,89 +60,99 @@ type Reconciler struct {
+		logging.FromContext(ctx).Desugar().Error("PubSub in work queue no longer exists")
 		return nil
 	} else if err != nil {
 		return err
 	}
 
 	// Don't modify the informers copy
-	csr := original.DeepCopy()
+	pubsub := original.DeepCopy()
+
+	reconcileErr := r.reconcile(ctx, pubsub)
 
-	reconcileErr := r.reconcile(ctx, csr)
+	// If no error is returned, mark the observed generation.
+	if reconcileErr == nil {
+		pubsub.Status.ObservedGeneration = pubsub.Generation
+	}
 
-	if equality.Semantic.DeepEqual(original.Status, csr.Status) &&
-		equality.Semantic.DeepEqual(original.ObjectMeta, csr.ObjectMeta) {
+	if equality.Semantic.DeepEqual(original.Status, pubsub.Status) {
 		// If we didn't change anything then don't call updateStatus.
 		// This is important because the copy we loaded from the informer's
 		// cache may be stale and we don't want to overwrite a prior update
 		// to status with this stale state.
-	} else if _, err := r.updateStatus(ctx, csr); err != nil {
-		// TODO: record the event (c.Recorder.Eventf(...
-		r.Logger.Warn("Failed to update PubSub Source status", zap.Error(err))
-		return err
-	}
 
+	} else if _, uErr := r.updateStatus(ctx, pubsub); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update PubSub status", zap.Error(uErr))
+		r.Recorder.Eventf(pubsub, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for PubSub %q: %v", pubsub.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(pubsub, corev1.EventTypeNormal, "Updated", "Updated PubSub %q", pubsub.Name)
+	}
 	if reconcileErr != nil {
-		// TODO: record the event (c.Recorder.Eventf(...
-		return reconcileErr
+		r.Recorder.Event(pubsub, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
 	}
-
-	return nil
+	return reconcileErr
 }
 
-func (r *Reconciler) reconcile(ctx context.Context, source *v1alpha1.PubSub) error {
-	source.Status.ObservedGeneration = source.Generation
-	source.Status.InitializeConditions()
+func (r *Reconciler) reconcile(ctx context.Context, pubsub *v1alpha1.PubSub) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("pubsub", pubsub)))
+
+	pubsub.Status.InitializeConditions()
 
-	if source.DeletionTimestamp != nil {
+	if pubsub.DeletionTimestamp != nil {
 		// No finalizer needed, the pullsubscription will be garbage collected.
 		return nil
 	}
 
-	ps, err := r.reconcilePullSubscription(ctx, source)
+	ps, err := r.reconcilePullSubscription(ctx, pubsub)
 	if err != nil {
-		r.Logger.Infof("Failed to reconcile PubSub: %s", err)
+		pubsub.Status.MarkPullSubscriptionNotReady("PullSubscriptionCreateFailed", "Failed to reconcile PullSubscription: %s", err.Error())
 		return err
 	}
-	source.Status.PropagatePullSubscriptionStatus(ps.Status.GetCondition(apis.ConditionReady))
+	pubsub.Status.PropagatePullSubscriptionStatus(ps.Status.GetCondition(apis.ConditionReady))
```

Comment:
Should this only be executed in the success case? Does this overwrite the reason set in the error case at L127?

Reply:
Yes. This propagation will either mark the Subscription ready or not. If the status is not ready, then it will overwrite the reason above.

Reply:
Ignore this thread, I missed that the branch at L127 returns before this line is hit.